### PR TITLE
Unify and clean up the common build properties

### LINKED ISSUE #922: Adjust the link line for all exe/dlls to prefer onecore_apiset
dwmapi, winmm, shlwapi are all in different forms on OneCore editions than on desktop, so linking those libs explicitly (instead of via umbrella) is probably not good.

pathcch isn't there at all... so maybe the API we're using through that is available in a different way?

_Originally posted by @miniksa in https://github.com/microsoft/terminal/pull/913_

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
## Summary of the Pull Request

This commit cleans up and deduplicates all of the common build
preamble/postamble across exe, dll, lib and c++/winrt projects.

The following specific changes have been made:
* All projects now define their ConfigurationType
* All projects now set all their properties *before* including a common
  build file (or any other build files)
* cppwinrt.pre and cppwinrt.post now delegate most of their
  configuration to common.pre and common.post
* (becuase of the above,) all build options are conserved between
  console and c++/winrt components, including specific warnings and
  preprocessor definitions.
* More properties that are configurable per-project are now
  conditioned so the common props don't override them.
* The exe, dll, exe.or.dll, and lib postincludes have been merged into
  pre or post and switched based on condition as required
* Shared items (-shared, -common) are now explicitly vcxitems instead of
  vcxproj files.
* The link line is now manipulated after Microsoft.Cpp sets it, so the
  libraries we specify "win". All console things link first against
  onecore_apiset.lib.
* Continues the tradition of naming PDB files `xxxFullPDB`, but I can't tell why we did that so this may be cargo cultism (I had to make the Cascadia projects follow that rule)

Fixes #922.

## PR Checklist
* [x] Closes #922
* [x] CLA signed.
* [x] Tests added/passed/project built a bunch
* [ ] Requires documentation to be updated
* [x] Core contributor

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
I forgot to go the vcxitems part. Following up. :)

File: src/cascadia/TerminalControl/UiaTextRange.cpp [hunk 1]
```diff
@@ -34,6 +34,7 @@ std::deque<UiaTextRange*> UiaTextRange::GetSelectionRanges(_In_ IUiaData* pData,
             {
                 UiaTextRangeBase* temp = ranges[0];
                 ranges.pop_front();
+                temp->Release();
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/UiaTextRange.cpp]: Where'd this come from?

File: src/common.build.pre.props [hunk 2]
```diff
@@ -63,8 +72,8 @@
       <!--<WarningLevel>EnableAllWarnings</WarningLevel>-->
       <TreatWarningAsError>true</TreatWarningAsError>
       <!-- disable warning on nameless structs (4201) -->
-      <DisableSpecificWarnings>4201;4312;4467</DisableSpecificWarnings>
-      <PreprocessorDefinitions>EXTERNAL_BUILD;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <DisableSpecificWarnings>4201;4312;4467;%(DisableSpecificWarnings)</DisableSpecificWarnings>
+      <PreprocessorDefinitions>_WINDOWS;EXTERNAL_BUILD;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Comment (zadjii-msft) [inline src/common.build.pre.props]: What does `_WINDOWS` do?

File: src/cppwinrt.build.pre.props [hunk 3]
```diff
@@ -43,97 +41,51 @@
...
+  <ItemDefinitionGroup>
     <ClCompile>
-      <IntrinsicFunctions>true</IntrinsicFunctions>
+      <AdditionalIncludeDirectories>$(OpenConsoleDir)\src\cascadia\WinRTUtils\inc;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
     </ClCompile>
     <Link>
-      <EnableCOMDATFolding>true</EnableCOMDATFolding>
-      <OptimizeReferences>true</OptimizeReferences>
+      <AdditionalDependencies>WindowsApp.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AppContainer>false</AppContainer>
```
Comment (zadjii-msft) [inline src/cppwinrt.build.pre.props]: what the heck is this

File: src/cascadia/TerminalControl/UiaTextRange.cpp (hunk 1 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalControl/UiaTextRange.cpp]: this was a code correctness issue that the new “unused locals” warning caught. Ranges is an unprotected vector of raw pointer types, so ref counting wasn’t being done on it.

File: src/common.build.pre.props (hunk 2 above)
Comment (DHowett-MSFT) [reply src/common.build.pre.props]: No idea, but we had it elsewhere so I merged it into here.

File: src/cppwinrt.build.pre.props (hunk 3 above)
Comment (DHowett-MSFT) [reply src/cppwinrt.build.pre.props]: This one turns off the app container bit in the final DLL or EXE. That bit is part of the reason we failed WACK with our first submission :)

**Comment by miniksa** [comment]
@DHowett-MSFT, are you 'vcxitems'ing the resources here too? I had that on my todo for the Universal variant.

**Comment by DHowett-MSFT** [comment]
@miniksa I wasn’t intending on it, but I can :)

**Comment by miniksa** [comment]
The one comment I'd like you to follow up on eventually is:
"Please make a docs\MSBuildWizardry.md with all the rules and nuances here."

I feel like we're doing a lot of custom or interesting MSBuild stuff which should be documented somehow like:
1. Moving the properties up to the top above the pre props so we can selectively pick them out in the pre files
2. List out what's supposed to go in pre (properties) and post (item groups) and when exceptions are OK (when stuff like link dependencies has to be in a specific spot)
3. Talk about the nuance of why most things are .vcxitems for just "grab bags" except the one that the WAP project references which is just .items (because the WAP project throws a fit and it is doing a simple extension match and will ignore it as just .items).

And so on and so forth. Maybe also include a link to that log debugging tool you used and how to generate a binlog.

**Pull request merged**